identity-core/src/common/one_or_set.rs

```diff
@@ -0,0 +1,484 @@
+use core::iter;
+use core::mem::replace;
+use core::ops::Deref;
+use core::slice::from_ref;
+
+use serde::de;
+use serde::Deserialize;
+use serde::Serialize;
+
+use identity_diff::Diff;
+use identity_diff::DiffVec;
+
+use crate::common::KeyComparable;
+use crate::common::OrderedSet;
+use crate::error::Error;
+use crate::error::Result;
+
+/// A generic container that stores exactly one or more unique instances of a given type.
+///
+/// Similar to [`OneOrMany`](crate::common::OneOrMany) except instances are guaranteed to be unique,
+/// and only immutable references are allowed.
+#[derive(Clone, Hash, PartialEq, Eq, PartialOrd, Ord, Deserialize, Serialize)]
+#[serde(transparent)]
+pub struct OneOrSet<T>(OneOrSetInner<T>)
+where
+  T: KeyComparable;
+
+// Private to prevent creations of empty `Set` variants.
+#[derive(Clone, Hash, PartialEq, Eq, PartialOrd, Ord, Deserialize, Serialize)]
+#[serde(untagged)]
+enum OneOrSetInner<T>
+where
+  T: KeyComparable,
+{
+  /// A single instance of `T`.
+  One(T),
+  /// Multiple (one or more) unique instances of `T`.
+  #[serde(deserialize_with = "deserialize_non_empty_set")]
+  Set(OrderedSet<T>),
+}
+
+/// Deserializes an [`OrderedSet`] while enforcing that it is non-empty.
+fn deserialize_non_empty_set<'de, D, T: serde::Deserialize<'de> + KeyComparable>(
+  deserializer: D,
+) -> Result<OrderedSet<T>, D::Error>
+where
+  D: de::Deserializer<'de>,
+{
+  let set: OrderedSet<T> = OrderedSet::deserialize(deserializer)?;
+  if set.is_empty() {
+    return Err(de::Error::custom(Error::OneOrSetEmpty));
+  }
+
+  Ok(set)
+}
+
+impl<T> OneOrSet<T>
+where
+  T: KeyComparable,
+{
+  /// Constructs a new instance with a single item.
+  pub fn new_one(item: T) -> Self {
+    Self(OneOrSetInner::One(item))
+  }
+
+  /// Constructs a new instance from a set of unique items.
+  ///
+  /// Errors if the given set is empty.
+  pub fn new_set(set: OrderedSet<T>) -> Result<Self> {
+    if set.is_empty() {
+      return Err(Error::OneOrSetEmpty);
+    }
+    if set.len() == 1 {
+      Ok(Self::new_one(
+        set.into_vec().pop().expect("infallible OneOrSet new_set"),
+      ))
+    } else {
+      Ok(Self(OneOrSetInner::Set(set)))
+    }
+  }
+
+  /// Construct a [`OneOrSet<T>`] from a [`OneOrSet<U>`] where `U` can be converted to `T`.
+  ///
+  /// Workaround for lack of specialisation preventing a generic `From` implementation.
+  pub fn from<U>(other: OneOrSet<U>) -> Self
+  where
+    U: KeyComparable + Into<T>,
+  {
+    Self(match other.0 {
+      OneOrSetInner::One(item) => OneOrSetInner::One(item.into()),
+      OneOrSetInner::Set(set_u) => {
+        let set_t: OrderedSet<T> = set_u.into_vec().into_iter().map(Into::into).collect();
+        // Key equivalence could differ between U and T.
+        if set_t.len() == 1 {
+          OneOrSetInner::One(set_t.into_vec().pop().expect("OneOrSet::from infallible"))
+        } else {
+          OneOrSetInner::Set(set_t)
+        }
+      }
+    })
+  }
```

Comment:
I think the following would be an improvement. 
```
 /// Takes a closure and transforms the [`OneOrSet`] by calling that closure on each element.  
  pub fn map<S, F>(self, mut f: F) -> OneOrSet<S>
  where
    S: KeyComparable,
    F: FnMut(T) -> S,
  {
    OneOrSet(match self.0 {
      OneOrSetInner::One(item) => OneOrSetInner::One(f(item)),
      OneOrSetInner::Set(set_u) => {
        let set_t: OrderedSet<S> = set_u.into_vec().into_iter().map(f).collect();
        // Key equivalence could differ between U and T.
        if set_t.len() == 1 {
          OneOrSetInner::One(set_t.into_vec().pop().expect("OneOrSet::from infallible"))
        } else {
          OneOrSetInner::Set(set_t)
        }
      }
    })
  }
```

Reply:
I used the same `from<U: Into<T>>` approach in other places like DIDUrl but I think `map` might be better for a collection. Thanks.